src/app/content/utils/canonicalUrl.spec.ts

```diff
@@ -217,4 +218,76 @@ describe('getCanonicalURL', () => {
       title: expect.anything(),
     }));
   });
+
+  it('finds the deepest canonical page', async() => {
+    helpers.archiveLoader.mockBook(mockBook);
+    helpers.archiveLoader.mockBook(mockOtherBook);
+
+    const bookId = book.id;
+    const pageId = page.id;
+
+    CANONICAL_MAP[bookId] = [['testbook2-uuid', { [pageId]: 'testbook2-page' }]];
+    CANONICAL_MAP['testbook2-uuid'] = [['testbook3-uuid', {'testbook2-page': 'testbook3-page'}]];
+    CANONICAL_MAP['testbook3-uuid'] = [['testbook3-uuid', {}]];
+
+    const node = archiveUtils.findArchiveTreeNodeById(book.tree, pageId);
+    const spy = jest.spyOn(archiveUtils, 'findArchiveTreeNodeById')
+      .mockReturnValueOnce({ ...node, slug: 'testbook2-page' } as LinkedArchiveTreeSection)
+      .mockReturnValue({ ...node, slug: 'testbook3-page' } as LinkedArchiveTreeSection);
```

Comment:
i'm not sure about these `findArchiveTreeNodeById` mocks, is it possible to work with stuff that is in the test tree? or set up the mock test tree? 

i feel like we may rearrange this logic again, it would be nice for these tests to not be coupled to the implementation logic